### PR TITLE
feat: Branded Link Surveys

### LINKED ISSUE #2119: [FEATURE][🕹️] Branded Link Surveys
### Is your feature request related to a problem? Please describe.

**Issue**: 
People want to brand surveys with a logo that is always visible

### Describe the solution you'd like

**Solution**: 
We allow people to add a company logo to their product:

- They can adjust the background color
- They can toggle it on and off for link surveys

**Survey Editor UI**
Add a field to the top left corner which prompts users to upload their company logo:

image

image

On click, the upload mask opens which only allows .png and .jpeg files. 

**Logo Settings**

Once an image file was selected and uploaded, this modal opens:
image

1. Replace logo: Opens the upload modal again to replace logo
2. Remove logo: Replaces the preview with our standard file upload UI (find it in `packages/ui`). However, the logo file is not deleted on button click, this happens only on "Save". Otherwise "Cancel" cannot undo it.
3. Background color: Default #ffffff, affects the background color of the container holding the logo

The rest of the UI should be self-explanatory. On SAVE the logo is displayed accordingly in the top left corner of the link survey preview. 

On HOVER on the logo, we see an EDIT icon button which opens the same settings modal:

image

In the Styling Card in the survey settings, we add a "Hide logo" toggle if the product has a logo:

image

The toggle hides the logo in this specific survey :)

----------

**Product Settings > Look & Feel Page**
We also allow to add the logo on the Look & Feel page in the Product Settings. The card looks like this at first:

image

Once the user uploaded an image we see the editor:

image

After save, we can see the preview and the color but cannot make changes without hitting "Edit":

image

--------

In the `LinkSurvey.tsx` display the logo if there is one for the product and the survey styling has showLogo enabled. It should be displayed in the top left corner of the page. Make sure it works well responsively.

--------

**Data Model**

Please extend the `Product` model in prisma with a new JSON object `brand`:

```
"brand": {
  "logoUrl": "...",
  "bgColor": "#ffffff",
}
```

In the styling JSON in the `Survey` model we set the variable `showLogo` to true for all new surveys. If  the variable is not set, we don't display the logo. In the settings in the Styling card we can set it to false by toggling off.

---------

**Team Roles**
We have team roles with different rights. Only render the card in the look and feel settings as well as the "Add logo here" indicator in the survey editor when the logged in user is a Developer, Editor or Admin (so not a Viewer).

---------

**How we code at Formbricks**
1. To keep our code DRY, create the UI we need both in the survey editor and the settings page in the `packages/ui` folder and import it back into the Formbricks app to use it.
2. Upload functionality: We're using it in several place in the app, stick to the same process to keep it consistent.

### Describe alternatives you've considered

-

### Additional context

-

### PR DESCRIPTION
**Pull request by manish-singh-bisht**
## What does this PR do?

Enables users to add their company logo to link surveys.

**technical description**- user starts to edit their survey and is prompted with "Add logo here" for link surveys, on clicking ->upload modal shows up->user selects logo ->the logo is uploaded-> the logo setting modal open-> user can choose to save->on save a call is made to updateProductAction and logoUrl and bgColor are set in db -> on hovering a pencil icon is shown ->clicking it opens the logosetting modal again->user can replace/cance/save->in the survey setting "hide logo" button is shown only when there is a company logo otherwise not-> toggling it hides/shows the logo in link survey.->toggling changes the showLogo property of styling json in survey model. Look and feel settings also has a new section "Logo" ->here also user can set the company.logo

Flow 
SurveyEditor -> PreviewSurvey->AddLogoButton->LogoSettingModal->LogoSetting
EditLogo(/settings/lookandfeel/components/EditLogo.tsx) ->LogoSetting

Fixes # (issue)

Fixes #2119 

Loom Video: https://www.loom.com/share/3babf06734d94f9b9463a05798839d6c
-->

## How should this be tested?

1.click on Add Logo here button in the link survey at the top left corner.
2.select and png,jpeg image
3.select a background color if you need one,and click save.it will be shown in preview.
4.user can toggle on and off the hideLogo in survey setting.toggle on ,hides the logo.
5.publish and visit the link and the logo will be at top left corner.
6.in the look and feel setting of product, the logo can be changed from there as well. 

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: packages/ui/LogoSettingModal/actions.ts
Comment (ShubhamPalriwala) [inline packages/ui/LogoSettingModal/actions.ts:19]: Add a check to throw an error if the product does not exist! Then you can also get rid of the force checker in 
 `product!.teamId`

File: packages/ui/AddLogoButton/lib/uploadLogo.tsx
Comment (ShubhamPalriwala) [inline packages/ui/AddLogoButton/lib/uploadLogo.tsx:1]: We have a very similar functionality in `apps/web/app/(app)/environments/[environmentId]/settings/profile/lib.ts`, try reusing that to avoid duplicated code in the codebase! Also not sometginf for the ui folder :)

File: packages/ui/AddLogoButton/index.tsx
Comment (ShubhamPalriwala) [inline packages/ui/AddLogoButton/index.tsx:99]: Instead of a 
`          {!product?.brand?.logoUrl && (`
and 
`          {product?.brand?.logoUrl && (`

try using something a ternary operator like `product?.brand?.logoUrl ?  <true logic> : <false div>`

Will make the code more readable

File: packages/ui/AddLogoButton/index.tsx
Comment (ShubhamPalriwala) [inline packages/ui/AddLogoButton/index.tsx:28]: here as well try to make the interface name less generic

File: packages/ui/AddLogoButton/index.tsx
Comment (ShubhamPalriwala) [inline packages/ui/AddLogoButton/index.tsx:21]: Won't this always be present?

File: apps/web/app/(app)/environments/[environmentId]/surveys/components/PreviewSurvey.tsx
Comment (ShubhamPalriwala) [inline apps/web/app/(app)/environments/[environmentId]/surveys/components/PreviewSurvey.tsx:32]: These will always be present right?

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/StylingCard.tsx
Comment (ShubhamPalriwala) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/StylingCard.tsx:359]: Can we simplify this condition with a variable above to add to the readability?

**Review by ShubhamPalriwala — COMMENTED** [review]
Hey @manish-singh-bisht thanks a lot for the PR ❤️ Great turnaround time and it feels great to use this.

I've mentioned some minor changes that I think will help us get the PR ready for a merge below so please take a look 🙏🏼 Do reach out if you have any queries/doubts. 

1. Can we adjust the placeholder position for the "Add Logo here" in Preview to have a similar horizontal & vertical margins? Looks kind of out of place now: 
image

2. For the Logo Settings Modal:
image
Can we:
    -  The Modal does not look vertically center, is there some padding/empty divs present? Can we fix this?
    - The Image Preview box has a grey background fill in the mock up just like you did in the Background color section, so can you do that here as well please?
    - The `The logo will be updated for all surveys which have the logo setting enabled` does not have the same border radius as the above section and others. Also verify to have similar border radius for the modal and internal sections
    - Can you make the text bold/heavier the `Preview` & `Background Color`?
    - The Modal is too wide in my opinion, can we reduce that as well?

3. On saving the Modal after making changes, my survey type turns back to "In-App Survey". Can you debug why this is happening? Might have to do something with the survey state that you store.

4. On Mobile Preview in Survey Editor, the Add Logo button gets misplaced:
image

5. On adding a Background color, the Preview in the Logo Settings Card does not match the Preview on Survey Editor:
image
image

PS: Its an image that does not have a background. The Logo Settings preview should only fill with the background color as much as it would on the actual survey.

6. On hovering over the uploaded logo, we see the pointer cursor, but on hovering over the edit icon on it, the pointer cursor as well as the edit icon disappears/flickers.
Screenshot_2024-03-20-19-59-54_5760x1080


7. In the Look & Feel page, on removing the logo & clicking save, the user should then not see the option to edit background etc, this does not seem to happen for me currently.
https://www.loom.com/share/382b27b71a954f3494812f5e7fdedb04?sid=827cf1cc-a849-4a6f-9d7b-16c79cc0c46d
 
8. With a survey that has a background color/image, on replacing the logo, removes that from the preview.
https://www.loom.com/share/1db591244fc24ebd826d064a73d8e043?sid=24679005-0bc1-4d4e-adfc-4b1d30e84c69

The core functionality works cleanly 💯 & nice work with following the DB schema & defaults 💪🏼! Really looking forward to get this feature shipped!

File: apps/web/app/(app)/environments/[environmentId]/surveys/components/PreviewSurvey.tsx
Comment (manish-singh-bisht) [reply apps/web/app/(app)/environments/[environmentId]/surveys/components/PreviewSurvey.tsx:32]: PreviewSurvey.tsx is being used at other places as well ,like here **apps/web/app/(app)/environments/[environmentId]/surveys/templates/TemplateContainer.tsx**

i need these two only for my AddLogoHere file, that is why these two are optional.

File: packages/ui/AddLogoButton/index.tsx
Comment (manish-singh-bisht) [reply packages/ui/AddLogoButton/index.tsx:21]: type can be made permanent, but membershipRole to be made  permanent ,then it should be permanent from the PreviewSurvey file.and preview survey file is being used at other places and from those these are not being provided,thats why optional.

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (manish-singh-bisht) [reply packages/ui/LogoSettingModal/components/LogoSetting.tsx:86]: setLocalProduct again to be permanent needs to be prermanent from PreviewSurvey and other places where previewSurvey is being used are not passing this.

setImageUrl comes from addLogoHere file, but doesnt come from EditLogo file in lookandfeel/component folder.

File: packages/ui/AddLogoButton/lib/uploadLogo.tsx
Comment (manish-singh-bisht) [reply packages/ui/AddLogoButton/lib/uploadLogo.tsx:1]: yes i was looking forward to this remark, i was not really sure regarding this file creation and also importing it from profile section.

will import it from lib.ts.

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (manish-singh-bisht) [reply packages/ui/LogoSettingModal/components/LogoSetting.tsx:99]: by cleaner you mean to not use the arbitary value but a tailwind class like z-50 or something?

this is used to prevent any editing after save/cancel button is pressed in look and feel settings.

**Comment by manish-singh-bisht** [comment]
for this in point 5 "PS: Its an image that does not have a background. The Logo Settings preview should only fill with the background color as much as it would on the actual survey."

if the survey say has an animation but we want the logo to be in white background , if the logo has transparent background then the bgcolor covers the whole transparent background of the logo giving it the white color needed(if white is selected in logosettings),  and if the logo has a background beforehand with itself then a little border around the logo with that bgcolor is shown.

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (manish-singh-bisht) [reply packages/ui/LogoSettingModal/components/LogoSetting.tsx:130]: gives error on changing

change:  `onFileUpload={(url: string[]) => {`

**_error_**:  Type '(url: string[]) => void' is not assignable to type '(uploadedUrl: string[] | undefined) => void'.
  Types of parameters 'url' and 'uploadedUrl' are incompatible.
    Type 'string[] | undefined' is not assignable to type 'string[]'.
      Type 'undefined' is not assignable to type 'string[]'.ts(2322)
index.tsx(22, 3): The expected type comes from property 'onFileUpload' which is declared here on type 'IntrinsicAttributes & FileInputProps'

**Comment by manish-singh-bisht** [comment]
for 8th point ,  the localSurvey state in the surveyEditor.tsx gets back to the data saved in the db,when the logo is uploaded.Users will have to save the progress before uploading the logo.this could also be the reason for the 3rd point why it is returning to app-survey.

**Comment by manish-singh-bisht** [comment]
hey @mattinannt ,is there something wrong that i did, why was removed from asignees?

**Comment by ShubhamPalriwala** [comment]
Hey @manish-singh-bisht, thanks a lot for working on all the pending tasks 💪🏼 I just tested the functionality & it all seems to work nicely! Can you resolve the merge conflicts & then one thorough round of testing after which we'll be good to go! So please ping me once this is done from your end! Excited for this 😄

**Comment by manish-singh-bisht** [comment]
hey @ShubhamPalriwala looks good from my side. ready for review.

**Comment by manish-singh-bisht** [comment]
@ShubhamPalriwala  i forgot to change the naming in link survey. have corrected that , everything should work now.

**Review by ShubhamPalriwala — COMMENTED** [review]
Hey, thanks for handling the conflicts! Faced a couple of small nitpicks:

1. The color picker does not close when clicking outside 
image

2. On accessing the link survey, I get the following error:
image

**Comment by manish-singh-bisht** [comment]
the second one is new, i dont know why the first is happening

will look into it.

**Comment by manish-singh-bisht** [comment]
HEY @ShubhamPalriwala  can you confirm the second one without my changes, i mean like remove my changes and test it . thank you

**Comment by manish-singh-bisht** [comment]
hey @ShubhamPalriwala  the color picker issue is somewhere related to it being inside modal,  whenever colorpicker is inside modal it is not closing on clicking outside , but clicking outside of modal in the blur area colorpicker closes. i have tested this by adding colorpicker to createTeamModal and same behaviour of colorpicker, not closing on outside click , but clicking on blur area of modal does close the color picker.

i have no clue on how to solve this.

**Comment by ShubhamPalriwala** [comment]
Thanks for the fixes! The window object error is not from this PR, we'll take a look at that separately. The color picker modal closes when clicked back on the same color or outside the modal so should work for now!

**Comment by manish-singh-bisht** [comment]
is there anything else remaining from my side? @ShubhamPalriwala

**Comment by jobenjada** [comment]
hey! Looks great, I have pushed a few tweaks re styling / sizing / consistency.

Especially with the CSS I had to do a lot of cleaning up. You use custom values way to much which creates an inconsistent UI. I think I gave you this feedback for the oss.gg PR already 🤓

## To Do

One thing is missing: It's not possible to set the background to transparent. This is especially needed if the logo already comes on a background.

Please replace the background color picker with this component you find in many of the settings cards:

image

If the toggle is off, we set the background color to transparent.

Other than that, it looks really good - great work! :)

**Comment by manish-singh-bisht** [comment]
theres a bug,

**Comment by manish-singh-bisht** [comment]
@ShubhamPalriwala @jobenjada  now its ready from my side.

@jobenjada the latest css changes, i need them.

thank you

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (ShubhamPalriwala) [inline packages/ui/LogoSettingModal/components/LogoSetting.tsx:196]: Can you correct this `htmlFor` label?

File: packages/ui/AddLogoButton/index.tsx
Comment (ShubhamPalriwala) [inline packages/ui/AddLogoButton/index.tsx:110]: We'd like to support all image mimes so can you try changing this with `image/*`?

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (ShubhamPalriwala) [inline packages/ui/LogoSettingModal/components/LogoSetting.tsx:171]: We'd like to support all image mimes so can you try changing this with `image/*`?

File: packages/ui/LogoSettingModal/components/LogoSetting.tsx
Comment (ShubhamPalriwala) [inline packages/ui/LogoSettingModal/components/LogoSetting.tsx:143]: Could you allow `jpg` as well here?

**Review by ShubhamPalriwala — COMMENTED** [review]
Hey, thanks for making the changes! 🤓 I tested the transparent background functionality and faced a bug where:
1. The loader disappeared before the image actually loaded in the preview box
2. The background color got auto-selected as white even when I disabled it before uploading.
Attaching a Loom video explaining the same above:
https://www.loom.com/share/fda3bd95dd5947da9ad9935887c6198d?sid=3d255654-d7e8-4d4b-b137-8fab31238c8c

I also tried uploading a square png image and then the logo preview is different from the view on the actual survey! Can you fix the preview field accordingly irrespective of the image dimensions?

Preview:
Screenshot_2024-04-03-10-30-02_5760x1080

Actual:
Screenshot_2024-04-03-10-30-57_5760x1080

Do ping once you've made the changes or have any queries so that I can review it again!

File: packages/ui/AddLogoButton/index.tsx
Comment (manish-singh-bisht) [reply packages/ui/AddLogoButton/index.tsx:110]: well in the ticket it was just png and jpeg,  will change to all

**Comment by manish-singh-bisht** [comment]
hey @ShubhamPalriwala the loader issue, it should not happen because in the handleUpload in logosetting file, setLoading(false) is set in the finally block.

**Comment by ShubhamPalriwala** [comment]
Thanks for the changes @manish-singh-bisht! 🚀 works nicely now 😄

**Comment by jobenjada** [comment]
have made quite a few changes to simplify both UX and code, feel free to have a look.

 will make final review tomorrow before going to merge 🚀

File: packages/ui/LogoPreview/index.tsx
Comment (mattinannt) [inline packages/ui/LogoPreview/index.tsx:24]: please remove and use `product.styling` directly

**Review by mattinannt — CHANGES_REQUESTED** [review]

File: apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx
Comment (jobenjada) [reply apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx:14]: moved this to lib folder and updated both settings which use it

File: apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx
Comment (jobenjada) [reply apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx:73]: working with null now

File: apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx
Comment (jobenjada) [reply apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx:88]: done

File: apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx
Comment (jobenjada) [reply apps/web/app/(app)/environments/[environmentId]/settings/lookandfeel/components/EditLogo.tsx:112]: working with null

File: packages/types/product.ts
Comment (jobenjada) [reply packages/types/product.ts:57]: done

File: packages/ui/AdvancedOptionToggle/index.tsx
Comment (jobenjada) [reply packages/ui/AdvancedOptionToggle/index.tsx:27]: done

File: packages/ui/LogoPreview/index.tsx
Comment (jobenjada) [reply packages/ui/LogoPreview/index.tsx:18]: updated

**Comment by jobenjada** [comment]
refactored according to plan. curenttly styling is not applied but not sure if related to changes.

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/CardStylingSettings.tsx
Comment (jobenjada) [reply apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/CardStylingSettings.tsx]: solved

File: packages/types/product.ts
Comment (mattinannt) [inline packages/types/product.ts:38]: please rename to `TLogo`

**Review by mattinannt — APPROVED** [review]

**Pull request merged**
